frontend/src/metabase/query_builder/components/expressions/ExpressionEditorTextfield.jsx

```diff
@@ -385,17 +388,38 @@ export default class ExpressionEditorTextfield extends React.Component {
       },
     },
     {
-      name: "tab",
-      bindKey: { win: "Tab", mac: "Tab" },
+      name: "chooseSuggestion",
+      bindKey: null,
       exec: () => {
-        this.handleTab();
+        this.chooseSuggestion();
       },
     },
   ];
 
+  // Correctly bind Tab depending on whether suggestions are available or now
+  updateTabBinding() {
+    if (this.input.current) {
+      const { editor } = this.input.current;
+      const { suggestions } = this.state;
+      const tabBinding = editor.commands.commandKeyBinding.tab;
+      if (suggestions.length > 0) {
+        // Something to suggest? Tab is for choosing one of them
+        editor.commands.bindKey("Tab", editor.commands.byName.chooseSuggestion);
+      } else {
+        if (Array.isArray(tabBinding) && tabBinding.length > 1) {
+          // No more suggestions? Keep a single binding and remove the
+          // second one (added to choose a suggestion)
+          editor.commands.commandKeyBinding.tab = tabBinding.shift();
+        }
+      }
+    }
+  }
+
   render() {
     const { source, suggestions, errorMessage, isFocused } = this.state;
 
+    this.updateTabBinding();
```

Comment:
This is indeed an atrocious place to invoke the function, due to its mutation aspect.

However, the alternative isn't clean either, i.e. sprinkling the call every time `suggestions` in the component state changes.

Reply:
Have you tried separating out this logic into a functional component? For example:

```jsx
const SuggestionsUpdater = ({input, suggestions}) => {
  const { current: { editor } = {}} = input;
  const tabBinding = editor?.commands?.commandKeyBinding?.tab;
  if (editor) {
    if (suggestions.length > 0) {
      // Something to suggest? Tab is for choosing one of them
      editor.commands.bindKey("Tab", editor.commands.byName.chooseSuggestion);
    } else {
      if (Array.isArray(tabBinding) && tabBinding.length > 1) {
          // No more suggestions? Keep a single binding and remove the
          // second one (added to choose a suggestion)
          editor.commands.commandKeyBinding.tab = tabBinding.shift();
      }
    }
  }
  return (<></>);
}
```

That way react can handle the lifetimes when updating

Reply:
It turns out that I was wrong: there's no need to "sprinkle" the call, since the suggestions update can be done in one place. I've added a commit that solves this cleanly.